### DISCUSSION TITLE
Best solution for using SSG dependening on whether there is a user session

### DISCUSSION THREAD

**Discussion by dmmz** [vercel/next.js#14564] [Help]
Hi, I have a question for a specific case where URL serves a static page (from SSG) but part of the page depends whether the user has logged in or not.. e.g.: there is header that would show the user's avatar in case there is a user session, or  a 'Log in' link otherwise.

What would be the best way to handle that? I can think of 4 options

Option 1.  With `useEffect` hook, we would re-render the header after getting that information (e.g. from a cookie)

Option 2. We would have two different routes,  
  - one would serve the static page, 
  - another one,  with with same URL but adding a  paramater `session=1` would use SSR. 

 With `useEffect` we would check if there is a session or not and would redirect properly. Something [like this](https://github.com/vercel/next.js/discussions/11346#discussioncomment-2091) 

Option 3. We would generate the static page at build time but the URL would always go to a custom server, where:
  - if there is a user session we would render the page 
  - otherwise we would serve the static page. Similar to what is [proposed here](https://github.com/vercel/next.js/discussions/14532) 

Option 4. We wouldn't use SSG at all.

As I am writing I think that I can split this use case into two.  Let's say that page shows some listing results:
- First use case. We decide data should be static for both logged in and logged out users. then option 1 may seem the best.
- Second use case: We decide logged in users will see always  fresh data, then option 1 is likely to be discarded.

What option would you recommend and why? or maybe there is a better option I didn't think of?

Thanks in advance

**Comment by getellez** [comment]
Hey so what was your approach? I want to use ssg to render some content but render an additional content if the user paid his membership

**Reply by dmmz** [reply]
At the time I posted the question we were using nextjs with Express as a server. We went for option 3. Express had the information of whether the user was logged in or not. If it was not, we returned the SSG rendered page, otherwise we would do SSR. 
It wasn't the perfect solution because we didn't benefit of SSG performance for logged in users.

In your case, if the additional content is not user specific content you could go for option 2 (with to SSG versions of the page) or 3, otherwise option 1 or 3. 

Whatever approach you choose, keep an eye on core web vitals. That's something we missed in the beginning at that time
